### ISSUE TITLE
BCOO matrix constructor fails within a `jax.lax.fori_loop`

### ISSUE THREAD

**Issue by GeoffNN** [google/jax#7971]
I'm implementing a power iteration routine in jax, using sparse BCOO matrices. 

I get the following error when using `fori_loop`: 

```ValueError: Invalid sparse representation: got indices.shape=(), data.shape=(), sparse_shape=(10,)```

In my routine, I fill the power iteration matrix incrementally, following the sparsity pattern of the vector. In this code, it is filled from a known dense matrix, for simplicity.

Minimal reproduction code: https://colab.research.google.com/drive/11hTyCoX30e054MI5zKV0lPaE9666YmUW?usp=sharing

**Comment by jakevdp** [comment]
Thanks! It looks like the issue boils down to trying to create a dynamically-sized array. A simpler repro of basically the same issue is this:
```python
from jax import jit
from jax.experimental import sparse
import jax.numpy as jnp

def f(M):
  return sparse.BCOO.fromdense(M)

M = jnp.arange(4).reshape(2, 2)

f(M)
# BCOO(int32[2, 2], nse=3)

jit(f)(M)
# ConcretizationTypeError: Abstract tracer value encountered where concrete value is expected...
```
When we call `jit(f)(M)`, it is attempting to construct a BCOO representation of a matrix `M` whose values are traced... this means that there's no way to know the required number of specified entries in the underlying representation.

If you want to create a BCOO matrix from a traced quantity, you need to specify the size of the representation statically, which you can do with the `nse` (number of specified elements) argument. For example:
```python
def g(M):
  return sparse.BCOO.fromdense(M, nse=3)

jit(g)(M)
# BCOO(int32[2, 2], nse=3)
```
You'll have to set `nse` based on what you know about the matrix being created. This is a bit of a wart in JAX's sparse implementation that comes from JAX transforms requiring static shapes, and it may be able to be relaxed in the future if and when the ongoing dynamic shape work lands.

**Comment by GeoffNN** [comment]
This seems like it's true also for sums of sparse matrices?

```python

def add(a, b):
  return jsparse.BCOO.fromdense(a, nse=2) + b

a = np.arange(4).astype(float)
b = jsparse.BCOO((jnp.ones(1), jnp.array([[0]])),
                 shape=(4,))

add(a, b)
# BCOO(float32[4], nse=3)

jax.jit(add)(a,b)
# TypeError: Cannot concatenate arrays with shapes that differ in dimensions other than the one being concatenated: concatenating along dimension 0 for shapes (1, 2), (1, 1).
```

**Comment by jakevdp** [comment]
Ah, yeah I see what's going on there. It's unrelated to the previous issue: we transposed the underlying BCOO representation but didn't account for the transpose in this impl. I can fix that bug.

**Comment by jakevdp** [comment]
I don't entirely understand why this only errors with JIT, but I'll look into it

**Comment by GeoffNN** [comment]
Thanks!

**Comment by GeoffNN** [comment]
Related to the `fori_loop` usage:
```python
b = jsparse.BCOO.fromdense(jnp.zeros(5).at[0].set(1.))

def add_a(i, a):
  return a + b

jax.lax.fori_loop(
    0, 10,
    add_a,
    b
)
```

This code raises this error:
`UnfilteredStackTrace: TypeError: JAX only supports number and bool dtypes, got dtype object in array`

In this case, the iterates have the same sparsity structure so it should work?
In the general case, is there a way to constraint the `nse` of the iterates to impose a static shape?

**Comment by jakevdp** [comment]
Thanks - something in the scan machinery seems to be initializing a `BCOO` object with abstract values, which shouldn't be happening. I'll look into it some more.

**Comment by jakevdp** [comment]
Oh I see - the issue is that `add_b` needs to have outputs that match the inputs, and for sparse arrays with arbitrary sparsity patterns, the output of `c = a + b` has `c.nse = a.nse + b.nse`, so the size of the output arrays does not match the size of the input arrays.

**Comment by jakevdp** [comment]
In this case, `a` and `b` do have matching sparsity patterns, but for a traced function `add_b` those sparsity patterns are not static, so there's no way for the implementation of the function to know that.

**Comment by jakevdp** [comment]
It might be that we should disable addition between sparse matrices without matching sparsity patterns... I think this behavior is probably surprising even in the best of cases, because repeated addition can lead to very inefficient sparse representations.

**Comment by jakevdp** [comment]
This TODO is the core issue in writing this code such that it recognizes the shared sparsity pattern: https://github.com/google/jax/blob/1182e7aaf6e42833dca65d0ac8afc84cdba8a1b8/jax/experimental/sparse/transform.py#L413-L416

I don't have any idea how to solve it.

**Comment by GeoffNN** [comment]
Could we somehow specify the `nse` of the result of the sum ? Somehow, it would be nice to have the result coalesce if needed, ie if the result has an `nse` larger than what it specifies, it would coalesce and return the `nse` first elements of the coalesced matrix, as is done in the constructors. This would allow 1) static `nse` 2) to still have correct results when summing sparse enough matrices / matrices with the same (or close) sparsity patterns.

**Comment by jakevdp** [comment]
Yeah that's possible, but there's no public API for it yet. You can hack it with something like this:
```python
import jax.numpy as jnp                                                                                                  
from jax.experimental import sparse as jsparse                                                                           

x = jsparse.BCOO.fromdense(jnp.array([1, 0, 0, 0]))                                                                      
y = x + x                                                                                                             

# deduplicate entries
y = y._dedupe()                                                                                                          

# trim indices and data to the first nse=1 entries
y.indices = y.indices[:1]
y.data = y.data[:1]

# Now the output has nse=1
print(y.todense())                                                                                                      
[2 0 0 0]
```
Currently the only mechanism we have to do shared-sparsity operations (due to the fact that indices are in general traced) is within the `sparsify` transform; primitives know how to track the sparsity pattern and know if it's shared, e.g.
```
@jsparse.sparsify 
def f(x): 
  return (2 * x) + x
print(x.nse, f(x).nse)                                                                                                  
# 1 1
```
but this is broken for higher-order primitives like `scan`, as mentioned in the TODO referred to above.

**Comment by GeoffNN** [comment]
Thanks for the explanation!

**Comment by GeoffNN** [comment]
Just checking: this is due to precisely the fact that the sum is used, correct?
The following code works:

```
M = jsparse.BCOO.fromdense(np.array([[1, 0],
                                     [1, 1]]), n_batch=1)
a = jsparse.BCOO.fromdense(jnp.array([1, 0.]), nse=2)

def power(i, v):
  return jsparse.BCOO.fromdense(M @ v, nse=2)

astar = jax.lax.fori_loop(
    0, 10,
    power,
    a
)

print(astar.todense())
# [ 1. 10.]
```

**Comment by GeoffNN** [comment]
Thanks for the clarification! I understand your previous code better. This works too:

```python
M = jsparse.BCOO.fromdense(np.array([[1, 0, 0],
                                     [1, 1, 0],
                                     [0, 0, 0]]), n_batch=1)
a = jsparse.BCOO.fromdense(jnp.array([1, 0., 0]), nse=2)

def power(i, v):
  return jsparse.BCOO.fromdense(M @ v, nse=2)

def sum_with_nse(i, mat, other_mat, nse):
  result = mat + other_mat
  result = result._dedupe()
  result.indices = result.indices[:nse]
  result.data = result.data[:nse]
  return result

astar = jax.lax.fori_loop(
    0, 10,
    functools.partial(sum_with_nse, other_mat=a, nse=2),
    a
)

print(astar.todense())
# [11.  0.  0.]
```

**Comment by jakevdp** [comment]
Yeah, that looks good. The larger question here... can we figure out a good way to provide the `sum_with_nse` type functionality more easily? So far I've left the `dedupe` functionality private and undocumented, but it's obviously useful in some cases.

**Comment by GeoffNN** [comment]
It probably wouldn't be necessary to use `_dedupe` at each sum here. My use case (which I suppose might be common) is that `nse` is bounded by max memory on a machine. It's typically larger than needed, but occasionally a dedupe might be necessary to stay under it. 

This also means we would need an easy way to _pad_ BCOO matrices up to a given `nse`, perhaps in the constructor, even when we're not using `fromdense` ?

**Comment by GeoffNN** [comment]
What I'm getting at, I think, is that maybe all ops should conserve `nse` by default ? 
I think we would want this behavior by default:
```
(a+b).nse == max(a.nse, b.nse) 
(A @ v).nse == v.nse
(A @ B).nse == max(A.nse, B.nse)
```
where typically we would start with an `nse` which upper bounds the maximum `nse` we will allow in our procedure.

**Comment by jakevdp** [comment]
That's not really possible, though. For example:
```python
a = BCOO.fromdense(jnp.array([1, 1, 0, 0]), nse=2)
b = BCOO.fromdense(jnp.array([0, 0, 1, 1]), nse=2)
c = a + b
```
Without concrete knowledge of the indices, the only knowledge you have is `c.nse <= a.nse + b.nse`.

**Comment by GeoffNN** [comment]
You could force b_padded to have `b_padded == a.nse + b.nse`, and then `(a+b).nse == max(a.nse, b_padded.nse)`

**Comment by jakevdp** [comment]
Won't `b_padded.nse` always be larger than `a.nse` in this case?

**Comment by GeoffNN** [comment]
Yes, so it'll always be `b_padded.nse`, and basically returns us to the easy case where `nse` is known before hand

**Comment by jakevdp** [comment]
I did leave out one thing: we could bound `nse` at the size of the axis, if we're willing to add the expense of an additional call to `dedupe`

**Comment by jakevdp** [comment]
But `b_padded.nse = a.nse + b.nse`, so I think you're describing the current behavior?

**Comment by GeoffNN** [comment]
Not quite: I'd like to pad b beforehand

**Comment by GeoffNN** [comment]
Suppose I know a pessimistic upperbound of `a.nse + b.nse` and pad `b` to that

**Comment by jakevdp** [comment]
I'm not entirely understanding... so stepping back, `a.nse` and `b.nse` are known concretely. The location of those specified elements is not known, so we pad the result of the sum to have `c.nse = a.nse + b.nse`. That's the pessimistic uppper-bound that can hold the correct sum in all cases. There's no way to reduce that without (1) possibly throwing away some specified entries in corner cases, or (2) applying some knowledge of the overlap in specified elements.

In general we cannot do (2), and (1) is unacceptable because it may lead to incorrect results, so we're left with `c.nse = a.nse + b.nse`.

Which part of that do you propose changing?

**Comment by GeoffNN** [comment]
`a.nse + b.nse` is the pessimistic _tight_ upperbound which can hold the correct sum in all cases. We don't necessarily need a tight upper bound; if we can relax that, then we can have a non-data dependent upper bound on the resulting `nse`.

**Comment by jakevdp** [comment]
So it sounds like you're assuming (2), you have some prior knowledge that there is some amount of overlap?

**Comment by GeoffNN** [comment]
Not really -- rather what I'm saying is that `2 * (a.nse + b.nse)` is also valid. Or if a and b have shape (100 000,) but we know that each has nse ~ 100, then maybe setting `1000` as an upper bound is fine, and will not blow up memory

**Comment by jakevdp** [comment]
OK, agreed, you can pad it as much as you want. But why would you want to allocate more memory than necessary?

**Comment by GeoffNN** [comment]
That's probably the bigger issue with this, agreed.

**Comment by jakevdp** [comment]
Yeah, I think that's why I'm leaning toward raising `NotImplemented` in this case.
